simpleclient/src/main/java/io/prometheus/client/Summary.java

```diff
@@ -38,133 +40,231 @@
-    private DoubleAdder sum = new DoubleAdder();
 
-    static TimeProvider timeProvider = new TimeProvider();
+    @Override
+    protected Child newChild() {
+        if (this.shouldCalculateQuantiles()) {
+            if(this.stream == null){
+                this.stream = new CKMSStream(500, quantiles);
+            }
+            return new Child(this.stream);
+        } else {
+            return new Child();
+        }
+    }
+
     /**
-     * Observe the given amount.
+     * Represents an event being timed.
      */
-    public void observe(double amt) {
-      count.add(1);
-      sum.add(amt);
+    public static class Timer {
+        Child child;
+        long start;
+
+        private Timer(Child child) {
+            this.child = child;
+            start = Child.timeProvider.nanoTime();
+        }
+
+        /**
+         * Observe the amount of time in seconds since {@link Child#startTimer} was called.
+         *
+         * @return Measured duration in seconds since {@link Child#startTimer} was called.
+         */
+        public double observeDuration() {
+            double elapsed = (Child.timeProvider.nanoTime() - start) / NANOSECONDS_PER_SECOND;
+            child.observe(elapsed);
+            return elapsed;
+        }
     }
+
     /**
-     * Start a timer to track a duration.
+     * The value of a single Summary.
      * <p>
-     * Call {@link Timer#observeDuration} at the end of what you want to measure the duration of.
+     * <em>Warning:</em> References to a Child become invalid after using
+     * {@link SimpleCollector#remove} or {@link SimpleCollector#clear}.
      */
-    public Timer startTimer() {
-      return new Timer(this);
+    public static class Child {
+        public static class Value {
+            private double count;
+            private double sum;
+        }
+
+        Child() {
+
+        }
+
+        Child(Stream<Double> stream) {
+            this.stream = stream;
+            shouldCalculateQuant = true;
+        }
+
+        // Having these seperate leaves us open to races,
+        // however Prometheus as whole has other races
+        // that mean adding atomicity here wouldn't be useful.
+        // This should be reevaluated in the future.
+        private DoubleAdder count = new DoubleAdder();
+        private DoubleAdder sum = new DoubleAdder();
+
+        // Boolean flag that dictates whether a Child instance should save data for subsequent quantile
+        // calculations.
+        private boolean shouldCalculateQuant = false;
+        private Stream<Double> stream;
+
+        static TimeProvider timeProvider = new TimeProvider();
+
+        /**
+         * Observe the given amount.
+         */
+        public void observe(double amt) {
+            count.add(1);
+            sum.add(amt);
+            if (this.shouldCalculateQuant) {
+
+                //  TODO
+                //  Will need to look into replacing the requirement.
```

Comment:
If you write a *TODO* in one line you be able to see a convenient (and readable) summary in your IDE